extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java

```diff
@@ -51,23 +99,196 @@ CodecProviderBuildItem collectCodecProviders(CombinedIndexBuildItem indexBuildIt
                 .collect(Collectors.toList());
     }
 
+    /**
+     * Create a producer bean managing the lifecycle of the MongoClient.
+     * <p>
+     * The generated class will look like
+     * 
+     * <pre>
+     * public class myclass extends AbstractMongoClientProducer {
+     *     &#64;ApplicationScoped
+     *     &#64;Produces
+     *     &#64;Default
+     *     public MongoClient createDefaultMongoClient() {
+     *         MongoClientConfig cfg = getDefaultMongoClientConfig();
+     *         return createMongoClient(cfg);
+     *     }
+     * 
+     *     &#64;ApplicationScoped
+     *     &#64;Produces
+     *     &#64;Default
+     *     public ReactiveMongoClient createDefaultReactiveMongoClient() {
+     *         MongoClientConfig cfg = getDefaultMongoClientConfig();
+     *         return createReactiveMongoClient(cfg);
+     *     }
+     * 
+     *     // for each named mongoclient configuration
+     *     // example:
+     *     // quarkus.mongodb.cluster1.connection-string = mongodb://mongo1:27017,mongo2:27017
+     *     &#64;ApplicationScoped
+     *     &#64;Produces
+     *     &#64;Named("cluster1")
+     *     &#64;MongoClientConnection("cluster1")
+     *     public MongoClient createNamedMongoClient_123456() {
+     *         MongoClientConfig cfg = getMongoClientConfig("cluster1");
+     *         return createMongoClient(cfg);
+     *     }
+     * 
+     *     &#64;ApplicationScoped
+     *     &#64;Produces
+     *     &#64;Named("cluster1")
+     *     &#64;MongoClientConnection("cluster1")
+     *     public ReactiveMongoClient createNamedReactiveMongoClient_123456() {
+     *         MongoClientConfig cfg = getMongoClientConfig("cluster1");
+     *         return createReactiveMongoClient(cfg);
+     *     }
+     * }
+     * </pre>
+     */
+    private void createMongoClientProducerBean(ApplicationArchivesBuildItem applicationArchivesBuildItem,
+            BuildProducer<GeneratedBeanBuildItem> generatedBean,
+            String mongoClientProducerClassName) {
+
+        Set<String> mongoClientNames = new HashSet<>();
+        IndexView indexView = applicationArchivesBuildItem.getRootArchive().getIndex();
+        Collection<AnnotationInstance> mongoClientAnnotations = indexView.getAnnotations(MONGOCLIENT_ANNOTATION);
+        for (AnnotationInstance annotation : mongoClientAnnotations) {
+            String mongoClientName = annotation.value().asString();
+            mongoClientNames.add(mongoClientName);
+        }
+
+        ClassOutput classOutput = new ClassOutput() {
+            @Override
+            public void write(String name, byte[] data) {
+                generatedBean.produce(new GeneratedBeanBuildItem(name, data));
+            }
+        };
+
+        ClassCreator classCreator = ClassCreator.builder().classOutput(classOutput)
```

Comment:
We usually wrap these calls in `try-with-resources` statements in order to avoid having to manually calls close at the end (of the potentially long) creation logic.
Same goes for the methods as well.

Reply:
done